File: src/VisualBasic/CodeCracker/Design/StaticConstructorExceptionAnalyzer.vb
Imports System.Collections.Immutable
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic

Namespace Design
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class StaticConstructorExceptionAnalyzer
        Inherits DiagnosticAnalyzer

        Public Shared ReadOnly Id As String = DiagnosticId.StaticConstructorException.ToDiagnosticId()
        Public Const Title As String = "Don't throw exceptions inside static constructors."
        Public Const MessageFormat As String = "Don't throw exceptions inside static constructors."
        Public Const Category As String = SupportedCategories.Design
    Public Const Description As String = "Static constructors are called before a class is used for the first time. Exceptions thrown
    in static constructors force the use of a try block and should be avoided."
        Protected Shared Rule As DiagnosticDescriptor = New DiagnosticDescriptor(
                Id,
                Title,
                MessageFormat,
                Category,
                SeverityConfigurations.CurrentVB(DiagnosticId.StaticConstructorException),
                isEnabledByDefault:=True,
                description:=Description,
                helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.StaticConstructorException))

        Public Overrides ReadOnly Property SupportedDiagnostics() As ImmutableArray(Of DiagnosticDescriptor) = ImmutableArray.Create(Rule)

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf Analyzer, SyntaxKind.SubNewStatement)
        End Sub

        Private Sub Analyzer(context As SyntaxNodeAnalysisContext)
            If (context.IsGenerated()) Then Return
            Dim ctor = DirectCast(context.Node, Syntax.SubNewStatementSyntax)
            If Not ctor.Modifiers.Any(SyntaxKind.SharedKeyword) Then Exit Sub

            Dim constructorBlock = DirectCast(ctor.Parent, Syntax.ConstructorBlockSyntax)
            If Not constructorBlock.Statements.Any() Then Exit Sub

            Dim throwBlock = constructorBlock.ChildNodes.OfType(Of Syntax.ThrowStatementSyntax).FirstOrDefault()
            If throwBlock Is Nothing Then Exit Sub

            context.ReportDiagnostic(Diagnostic.Create(Rule, throwBlock.GetLocation, Title))
        End Sub
    End Class
End Namespace

File: src/CSharp/CodeCracker/Design/StaticConstructorExceptionAnalyzer.cs
using CodeCracker;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Immutable;
using System.Linq;

namespace CodeCracker.CSharp.Design
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class StaticConstructorExceptionAnalyzer : DiagnosticAnalyzer
    {
        internal const string Title = "Don't throw exceptions inside static constructors.";
        internal const string MessageFormat = "Don't throw exceptions inside static constructors.";
        internal const string Category = SupportedCategories.Design;
        const string Description = "Static constructors are called before a class is used for the first time. Exceptions thrown "
            + "in static constructors force the use of a try block and should be avoided.";

        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
            DiagnosticId.StaticConstructorException.ToDiagnosticId(),
            Title,
            MessageFormat,
            Category,
            SeverityConfigurations.CurrentCS[DiagnosticId.StaticConstructorException],
            true,
            description: Description,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.StaticConstructorException));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);

        public override void Initialize(AnalysisContext context) =>
            context.RegisterSyntaxNodeAction(Analyzer, SyntaxKind.ConstructorDeclaration);

        private static void Analyzer(SyntaxNodeAnalysisContext context)
        {
            if (context.IsGenerated()) return;
            var ctor = (ConstructorDeclarationSyntax)context.Node;

            if (!ctor.Modifiers.Any(SyntaxKind.StaticKeyword)) return;

            if (ctor.Body == null) return;

            var @throw = ctor.Body.ChildNodes().OfType<ThrowStatementSyntax>().FirstOrDefault();

            if (@throw == null) return;

            context.ReportDiagnostic(Diagnostic.Create(Rule, @throw.GetLocation(), ctor.Identifier.Text));
        }
    }
}
